File: trees/BST.java
public class BST<Key extends Comparable<Key>, Val> {

    private Node root;

    private class Node {
        private Key key;
        private Val val;
        private Node left;
        private Node right;

        private Node(Key key, Val val) {
            this.key = key;
            this.val = val;
        }
    }

    public BST() {
        root = null;
    }

    public void put(Key key, Val val) {
        root = put(root, key, val);
    }

    private Node put(Node x, Key key, Val val) {

        if (x == null) {
            return new Node(key, val);
        }

        int cmp = key.compareTo(x.key);
        if (cmp < 0) {
            x.left = put(x.left, key, val);
        }
        else if (cmp > 0) {
            x.right = put(x.right, key, val);
        }
        else {
            x.val = val;
        }
        return x;
    }

    public Val get(Key key) {

        return get(root, key);
    }

    private Val get(Node x, Key key) {

        if (x == null) return null;

        int cmp = key.compareTo(x.key);
        if (cmp < 0) {
            return get(x.left, key);
        }
        else if (cmp > 0) {
            return get(x.right, key);
        }
        else
            return x.val;
    }

    public static void main(String[] args) {

        BST<Integer, String> bst = new BST<>();
        bst.put(1, "a");
        bst.put(2, "b");
        bst.put(3, "c");
        bst.put(4, "d");
        bst.put(5, "e");
        bst.put(6, "f");

        System.out.println(bst.get(4));
    }
}

File: trees/binary_search_tree/Python/BST.py
#!/usr/bin/env python
# coding: utf-8

# In[5]:


class Node: 
    def __init__(self,key): 
        self.left = None
        self.right = None
        self.val = key 

# insert a new node with the given key 
def insert(root,node): 
    if root is None: 
        root = node 
    else: 
        if root.val < node.val: 
            if root.right is None: 
                root.right = node 
            else: 
                insert(root.right, node) 
        else: 
            if root.left is None: 
                root.left = node 
            else: 
                insert(root.left, node) 

# inorder tree traversal 
def inorder(root): 
    if root: 
        inorder(root.left) 
        print(root.val) 
        inorder(root.right) 
        

#Driver code  
r = Node(5) 
insert(r,Node(3)) 
insert(r,Node(2)) 
insert(r,Node(4)) 
insert(r,Node(7)) 
insert(r,Node(6)) 
insert(r,Node(8)) 
  
# Print inoder traversal of the BST 
inorder(r) 


# In[ ]:




